plugin/pulpcore/plugin/download/file.py

```diff
@@ -0,0 +1,29 @@
+import errno
+
+from django.core.files import File
+
+
+class DownloadedFile(File):
```

Comment:
This is not part of the `File` interface.  We should not be doing special things so that a specific `Storage` will work correctly.

Reply:
I borrowed this code from the TemporaryUploadedFile[0]. The FileSystemStorage is designed to work with 2 types of uploaded files: In memory or on disk. Our upload API always writes files to disk. So when we hand an uploaded file to the FileSystemStorage, the file system storage calls file_move_safe(). I would like us to handle temporary downloaded files in the same manner.



[0] https://github.com/django/django/blob/1.11/django/core/files/uploadedfile.py#L59

Reply:
The original (code you removed) FileSystem._save() handled both in memory and on disk files with very little complexity and code.  It cleanly supported both linking and copying.  And, did not need the `content` to be a special type of `File` to do so.  The `FileSystemStorage._save()` seems over engineered and does not behave like pulp needs it to.  I don't see the point to writing code to coerce FileSystemStorage._save() to behave the way we want instead of just overriding _save(), path() and get_available_name().

After reviewing the `Storage` and `FileSystemStorage` code (again),  I still think the original FileSystem.path() and FileSystem.get_available_name() overrides that returned *name* was the correct thing to do.

I'm all for leveraging django code as much as possible so long as it's safe, practical and provides desirable functionality.